### PR TITLE
Improve TOTP field

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Fix misaligned TOTP field icon, make password field larger when TOTP is enabled and fix auto-submission to only fire when the user has already put in something into the password field.

# Screenshots

### TOTP not enabled, before and after
Screenshot from 2023-10-30 08-18-27

### TOTP enabled, `development-v6`
(misaligned icon, small password field)
Screenshot from 2023-10-30 08-19-10

### TOTP enabled, this PR
Screenshot from 2023-10-30 08-27-49

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Comment by DL6ER** [comment]
https://web.dev/articles/sign-in-form-best-practices

This adds a "show password" button:
image

**Review by yubiuser — COMMENTED** [review]
I think we should add a bit of space between the `https` warning and `show password`
Screenshot at 2023-10-30 20-40-06

**Comment by rdwebdesign** [comment]
Do we need the key icon?
Can we use an eye icon as "show password" button?

Something like this:
password_reveal

**Comment by DL6ER** [comment]
That's a very good solution, if you have this already at hand ;-) Feel free to push ti this branch

**Comment by rdwebdesign** [comment]
I will do it in a few minutes.

**Review by yubiuser — COMMENTED** [review]
The click/hover animation (color) changes stays after click until one clicks somewhere else.
Peek 2023-11-03 22-13

I think this should be a hover only.

**Review by yubiuser — COMMENTED** [review]
Entering the password first and submit correct TOTP afterwards shows "Wrong password" briefly before auto-login

Peek 2023-11-03 22-20

File: login.lp [hunk 1]
```diff
@@ -43,21 +43,20 @@ mg.include('scripts/pi-hole/lua/header.lp','r')
...
+                            </span>
                         </div>
                     </div>
-                    <!--
-                    <div class="form-group" title="Pi-hole has to set a cookie for the login session to be successful. The cookie will not contain your password nor is it used anywhere outside of you local Pi-hole.">
-                        <input type="checkbox" id="logincookie" checked>
-                        <label for="logincookie">Keep me logged in (uses cookie)</label>
-                    </div> -->
+                    <div class="form-group has-feedback hidden" id="totp_input">
+                        <input type="text" id="totp" size="6" maxlen="6" class="form-control totp_token" placeholder="123456" value="" spellcheck="false" autofocus>
```
Comment (yubiuser) [inline login.lp]:
```suggestion
                        <input type="text" id="totp" size="6" maxlen="6" class="form-control totp_token" placeholder="123456" value="" spellcheck="false" autocomplete="off" autofocus>
```

Not sure if this is the right command, but there should be no (sensless) autofill/complete on the TOTP input field

**Review by yubiuser — COMMENTED** [review]
There is no distinction between entering a wrong password or a wrong TOTP. It will always complain about "Wrong password" and autofocus on the password field - even when the TOTP was wrong.

**Comment by rdwebdesign** [comment]
This is the `:focus` style.
Every button has the same effect, but usually the focus changes to some other element after the click.

We can change the focus to the password field, like this:
button_focus

**Comment by DL6ER** [comment]
> Entering the password first and submit correct TOTP afterwards shows "Wrong password" briefly before auto-login

I am not able to reproduce this and looking again at the Javascript code, it also isn't immediately obvious to me how this could be happening. Could you try again and screenshot the `Network` tab frpm your browser for me with additional screenshots of the reply contents?

Please make sure to have "Persist Logs" enabled for them to "survive" the login process.

image

---

> There is no distinction between entering a wrong password or a wrong TOTP. It will always complain about "Wrong password" and autofocus on the password field - even when the TOTP was wrong.

I will change this to be clearer.

**Comment by DL6ER** [comment]
## Wrong password

Screenshot from 2023-11-04 07-16-01

## Correct password but wrong 2FA token

Screenshot from 2023-11-04 07-16-17

**Comment by yubiuser** [comment]
If you think letting users know if they entered a wrong password or a wrong TOTP is a security issue, I'm fine with leaving it as it is. I know this discussion, but it feels a bit like "security by obfuscation". If the second factor is so weak that it can be brute-forced within 30 seconds (before a new TOTP is generated), the second factor is broken anyway.

**Comment by DL6ER** [comment]
Yeah, since online-bruteforcing is prevented by FTL and when they can get physical/`root` access to either the Pi-hole or the device having the Authenticator app they'll trivially have the TOTP secret, I am fine with showing the user whether it was the password or the token that was wrong. Seem more user-friendly to me.

**Comment by yubiuser** [comment]
> Entering the password first and submit correct TOTP afterwards shows "Wrong password" briefly before auto-login

I'm not seeing this today, but there are two calls to `/auth` with the second one failing which might be the reason I saw it yesterday. Disabling 2FA, only one `/auth` request is logged.

Screenshot at 2023-11-05 20-47-49

**Comment by yubiuser** [comment]
The issue observed was triggered by `pasting` via Ctl+V  being recognized as two events (`paste` and `keyup`). I pushed a commit to only trigger on `input`. I tested it with pre-filled password with: entering TOTP digit by digit, Ctr-V and pasting with mouse. No double login was performed.

**Comment by yubiuser** [comment]
Is there a reason we prevent a TOTP re-use within 30 seconds?

```
2023-11-05 23:13:25.448 [399838/T404370] INFO: 2FA code verified successfully at 0
2023-11-05 23:13:33.616 [399838/T404372] WARNING: 2FA code has already been used (-1, 437220), please wait 27 seconds
2023-11-05 23:13:33.618 [399838/T404372] WARNING: API: Invalid 2FA token
```

This seems to be a too long lock interval.

**Comment by DL6ER** [comment]
> Is there a reason we prevent a TOTP re-use within 30 seconds?

It is a Time-Based ***One-Time Password*** Algorithm :-)

FTL has the answer to this question here:

> Verify code
> RFC 6238 (section 4.2): If the calculated value matches the value provided by the user, then the user is authenticated
> RFC 6238 (section 4.3): The server MUST NOT accept a TOTP value generated more than 30 seconds in the future
> RFC 6238 (section 4.3): The server MUST NOT accept a TOTP value generated more than 30 seconds in the past
> RFC 6238 (section 4.3): **The server MUST NOT accept a TOTP value it accepted previously**

furthermore:

> We RECOMMEND a default time-step size of 30 seconds.  This default value of 30 seconds is selected as a balance between security and usability.

and sections 5.3:

> Note that a prover may send the same OTP inside a given time-step window multiple times to a verifier.  The verifier **MUST NOT** accept the second attempt of the OTP after the successful validation has been issued for the first OTP, which ensures one-time only use of an OTP.


https://github.com/pi-hole/FTL/blob/afb6f865187099557aa6bb3332286ca2bf6730e7/src/api/2fa.c#L223-L224

The point is: Someone can somehow see your keystrokes. You enter your password. You enter your TOTP code. They immediately try to use them to log in and they'd succeed. The motivation behind the last point is that, even though they're still within the 30 second window for that code, it shouldn't work for they - specifically because you already used it.

**Comment by DL6ER** [comment]
With the most recent changes, the login page can much better react on errors:

## Initial state

Screenshot from 2023-11-06 11-05-15

## No password is supplied, submission is prevented

Screenshot from 2023-11-06 11-06-54

## Wrong password is entered

Screenshot from 2023-11-06 11-05-28

## Rate-limited
(simply mash <kbd>Enter</kbd>)
Screenshot from 2023-11-06 11-06-25

## Wrong 2FA token is provided

Screenshot from 2023-11-06 11-07-37

## Number of API seats is exceeded
(set `webserver.api.max_sessions = 1` and try to login in a private window)
Screenshot from 2023-11-06 11-09-14

---

For rate-limiting and API seat exhaustion to be reported correctly, you have to use FTL from https://github.com/pi-hole/FTL/pull/1733 (branch `new/num_sessions`)

**Comment by DL6ER** [comment]
The last commit adds the inclusion of `hint`s from FTL (if available). For example:

Screenshot from 2023-11-06 11-34-05

and

Screenshot from 2023-11-06 11-34-21

**Comment by rdwebdesign** [comment]
I noticed this wasn't reviewed after your last 2 commits.

I tested and it looks fine, but I'm requesting _yubiuser_ approval, because you 2 were discussing about error/warning messages.

**Review by yubiuser — APPROVED** [review]

**Comment by sefinek** [comment]
With the [LastPass](https://www.lastpass.com) extension, the `width` of the `input` is other.

https://github.com/pi-hole/web/assets/92880265/6d99823e-1291-4ea9-906c-01fa7f7757bb

**Comment by rdwebdesign** [comment]
This is an external bug.

LastPass extension modifies HTML page content (adding a `<div>` element to hold the icon) and their modification doesn't work here.
Not our fault, they just make assumptions they shouldn't make.

For example: 
if they had used `position: absolute` instead of relative, the layout wouldn't break here.
lastpass_issue

**Comment by sefinek** [comment]
Maybe it's worth informing them about it?

**Comment by yubiuser** [comment]
Sure, go ahead. You can link the issue/analysis here so they know what you are referring to.

**Pull request merged**